studio/studio.py

```diff
@@ -237,6 +245,171 @@ async def export(q: Q):
+                    ui.button('remove_package', label='Remove', value=package_name),
+                ]) for package_name, package_version in
+                [package.removesuffix('\n').split('==') for package in file.readlines()]
+            ],
+        ) if packages_installed else ui.text(name='no_packages', content='No packages installed.'),
+        ui.inline(
+            align='end',
+            justify='center',
+            height='60px',
+            items=[ui.buttons(items=[
+                ui.button(name='show_add_package_fields', label='Add package', icon='Add', primary=True, width='200px'),
+                ui.button(name='show_add_requirements', label='Add from requirements', icon='AddToShoppingList',
+                          width='200px')
+            ], justify='center')]
+        ),
+    ]
+
+
+def get_output(output: str) -> str:
+    return f'''
+```
+{output}
+```
+'''
+
+
+async def show_progress(q: Q, msg: str):
+    q.page['meta'].dialog.items = [
+        ui.progress(label=msg),
+        ui.expander(name='console_expander', label='Detail', items=[
+            ui.text(name='console_out', content=get_output(''), width='100%'),
+        ]),
+        ui.button(name='cancel_pip_task', label='Cancel')
+    ]
+    await q.page.save()
+
+
+async def update_progress(q: Q, value: int):
+    q.page['meta'].dialog.items[1].expander.items[0].text.content = get_output(value)
+    await q.page.save()
+
+
+async def show_finish_message(q: Q, type: Literal['error', 'success'], title: str, output: str):
+    q.page['meta'].dialog.items = [
+        ui.message_bar(type=type, text=title),
+        ui.expander(name='console_expander', label='Detail', items=[
+            ui.text(name='console_out', content=get_output(output), width='100%'),
+        ]),
+        ui.button(name='finish_message_dismiss', label='Go to package manager')
+    ]
+    await q.page.save()
+
+
+async def block_dialog(q: Q):
+    q.page['meta'].dialog.closable = False
+    q.page['meta'].dialog.blocking = True
+    await q.page.save()
+
+
+async def on_pip_finish(q: Q):
+    q.page['meta'].dialog.blocking = False
+    q.page['meta'].dialog.closable = True
+    q.page['meta'].dialog.items = get_package_dialog_items()
+    await q.page.save()
+
+
+async def pip(q: Q, command: str, flag: str or None, package_name: str, package_version: str, progress_message: str,
+              success_message: str, error_message: str, on_success: Callable = None, on_error: Callable = None):
+    try:
+        version = f'=={package_version}' if package_version else ''
+        args = [sys.executable, '-m', 'pip', command, flag, f'{package_name}{version}']
+        p = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output = ''
+        await block_dialog(q)
+        await show_progress(q, progress_message)
+        while True:
+            if p.poll() is not None:
```

Comment:
Doesn't this kill the main thread?

Reply:
The `subprocess.Popen()` is a non-blocking call. Using this kind of event loop inside its context should not be killing the main thread because if task is still running `p.poll()` returns `None` and UI is updated with `await update_progress(q, output)`. In case task is finished, `p.poll()` returns some value and the UI is updated accordingly as well.